tests/e2e/v3_curl_kv_test.go

```diff
@@ -16,86 +16,121 @@ package e2e
 
 import (
 	"encoding/json"
-	"path"
 	"testing"
 
+	"github.com/grpc-ecosystem/grpc-gateway/runtime"
+	"github.com/stretchr/testify/require"
+
 	pb "go.etcd.io/etcd/api/v3/etcdserverpb"
 	"go.etcd.io/etcd/pkg/v3/expect"
 	"go.etcd.io/etcd/tests/v3/framework/e2e"
-
-	"github.com/grpc-ecosystem/grpc-gateway/runtime"
 )
 
-func TestV3CurlPutGetNoTLS(t *testing.T) {
-	for _, p := range apiPrefix {
-		testCtl(t, testV3CurlPutGet, withApiPrefix(p), withCfg(*e2e.NewConfigNoTLS()))
-	}
+func TestCurlV3KVBasicOperation(t *testing.T) {
+	testCurlV3KV(t, testCurlV3KVBasicOperation)
 }
-func TestV3CurlPutGetAutoTLS(t *testing.T) {
-	for _, p := range apiPrefix {
-		testCtl(t, testV3CurlPutGet, withApiPrefix(p), withCfg(*e2e.NewConfigAutoTLS()))
-	}
-}
-func TestV3CurlPutGetAllTLS(t *testing.T) {
-	for _, p := range apiPrefix {
-		testCtl(t, testV3CurlPutGet, withApiPrefix(p), withCfg(*e2e.NewConfigTLS()))
-	}
+
+func TestCurlV3KVTxn(t *testing.T) {
+	testCurlV3KV(t, testCurlV3KVTxn)
 }
-func TestV3CurlPutGetPeerTLS(t *testing.T) {
-	for _, p := range apiPrefix {
-		testCtl(t, testV3CurlPutGet, withApiPrefix(p), withCfg(*e2e.NewConfigPeerTLS()))
-	}
+
+func TestCurlV3KVCompact(t *testing.T) {
+	testCurlV3KV(t, testCurlV3KVCompact)
 }
-func TestV3CurlPutGetClientTLS(t *testing.T) {
-	for _, p := range apiPrefix {
-		testCtl(t, testV3CurlPutGet, withApiPrefix(p), withCfg(*e2e.NewConfigClientTLS()))
+
+func testCurlV3KV(t *testing.T, f func(ctlCtx)) {
+	testCases := []struct {
+		name string
+		cfg  ctlOption
+	}{
+		{
+			name: "noTLS",
+			cfg:  withCfg(*e2e.NewConfigNoTLS()),
+		},
+		{
+			name: "autoTLS",
+			cfg:  withCfg(*e2e.NewConfigAutoTLS()),
+		},
+		{
+			name: "allTLS",
+			cfg:  withCfg(*e2e.NewConfigTLS()),
+		},
+		{
+			name: "peerTLS",
+			cfg:  withCfg(*e2e.NewConfigPeerTLS()),
+		},
+		{
+			name: "clientTLS",
+			cfg:  withCfg(*e2e.NewConfigClientTLS()),
+		},
 	}
-}
 
-func TestV3CurlTxn(t *testing.T) {
-	for _, p := range apiPrefix {
-		testCtl(t, testV3CurlTxn, withApiPrefix(p))
+	for _, tc := range testCases {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			testCtl(t, f, tc.cfg)
+		})
 	}
 }
 
-func testV3CurlPutGet(cx ctlCtx) {
+func testCurlV3KVBasicOperation(cx ctlCtx) {
```

Comment:
Using `ctlCtx` is the old and at least for me a deprecated way to write e2e test. Would appreciate if you considered a followup PR to refactor it.